### PR TITLE
Use `async-http-faraday`.

### LINKED ISSUE #774: Nondeterministic moving/deleting of PRs in CHANGELOG.md
I'm running this, with version `1.15.0`:

```
github_changelog_generator --user JacobEvelyn --project friends
```

and each time I run it the `CHANGELOG.md` changes—PRs that were listed under one of my gem's versions (say, `0.42`) will jump to a completely different version (say, `0.38`). When I run `github_changelog_generator` again they'll move back but different PRs will be switched around, or even deleted entirely!

I haven't seen this nondeterministic behavior before and have been on the same version of `github_changelog_generator` for a while.

EDIT: I just saw that this is fixed with #729, which has been open for almost a year.

### PR DESCRIPTION
**Pull request by ioquatix**
This is an alternative to https://github.com/github-changelog-generator/github-changelog-generator/pull/729

Be aware that until https://github.com/sourcelevel/faraday-http-cache/issues/118 is released, `async-http-faraday` is unable to use persistent connections correctly because that dependency pulls in an old release of faraday.

In addition, I had to add a monkey patch  because the Sawyer agent doesn't expose any way to close the underlying persistent connection, and we are waiting on https://github.com/lostisland/sawyer/pull/67 to be merged/released.

---

Fixes #774.

### CODE REVIEW

**Comment by ioquatix** [comment]
Also, there is no hope that this code will work on JRuby, until they fix their network layer: https://github.com/jruby/jruby/issues/4748

**Comment by ioquatix** [comment]
Even though performance can't be assessed accurately, here is some interesting information:

As a side note, I also found you have a shared file system cache. Has anyone checked if that is safe when multiple processes are running? I cleared it out between each run.

## master branch

```
Generated log placed in /home/samuel/Documents/ioquatix/github-changelog-generator/tmp/CHANGELOG.md
bundle exec ../bin/github_changelog_generator -u socketry -p async -t   11.82s user 0.42s system 42% cpu 28.541 total
```

## `fix-thread-unsafe` branch

```
Generated log placed in /home/samuel/Documents/ioquatix/github-changelog-generator/tmp/CHANGELOG.md
bundle exec ../bin/github_changelog_generator -u socketry -p async -t   11.38s user 0.46s system 41% cpu 28.467 total
```

## `async-http-faraday` branch

```
Generated log placed in /home/samuel/Documents/ioquatix/github-changelog-generator/tmp/CHANGELOG.md
bundle exec ../bin/github_changelog_generator -u socketry -p async -t   12.83s user 0.50s system 40% cpu 32.684 tota
```

I'll be more interested if we can use persistent connections, so I bypassed the cache:

## `async-http-faraday` without `faraday-http-cache`

```
Generated log placed in /home/samuel/Documents/ioquatix/github-changelog-generator/tmp/CHANGELOG.md
bundle exec ../bin/github_changelog_generator -u socketry -p async -t   10.34s user 0.63s system 47% cpu 23.075 total
```

It might be better to try it on a repo which has more history to really see whether there are significant improvements.

**Comment by ioquatix** [comment]
Ruby 2.3 and 2.4 are EOL, I guess that might include jruby 91?

**Comment by ioquatix** [comment]
I am now happy with this, we just need to rebase on https://github.com/github-changelog-generator/github-changelog-generator/pull/786 once it's merged.

**Comment by ioquatix** [comment]
Also, this branch is about 20-40% faster depending on what you are doing, and there is other low hanging fruit to improve perf but I don't have any more time for this right now, so once it's merged we can look at improving performance further.

**Comment by ioquatix** [comment]
Here is perf comparison:

### fix-thread-unsafe

```
Generated log placed in /home/samuel/Documents/ioquatix/github-changelog-generator/lib/CHANGELOG.md
bundle exec ../bin/github_changelog_generator --no-http-cache -u  -p  -t   25.64s user 0.88s system 26% cpu 1:39.57 total
```

### async-http-faraday

```
Generated log placed in /home/samuel/Documents/ioquatix/github-changelog-generator/lib/CHANGELOG.md
bundle exec ../bin/github_changelog_generator --no-http-cache -u  -p  -t   7.70s user 0.48s system 14% cpu 57.872 total
```

### changelogs

They are the same

```
koyoko% shasum CHANGELOG*
390b735e55a97578f597d0919eedf3e14f43ea07  CHANGELOG-async.md
390b735e55a97578f597d0919eedf3e14f43ea07  CHANGELOG-threads.md
```

**Comment by ioquatix** [comment]
Just for fun, here is master:

```
Generated log placed in /home/samuel/Documents/ioquatix/github-changelog-generator/tmp/CHANGELOG.md
bundle exec ../bin/github_changelog_generator -u github-changelog-generator -  26.77s user 1.08s system 30% cpu 1:31.09 total
koyoko% shasum *.md
390b735e55a97578f597d0919eedf3e14f43ea07  CHANGELOG-async.md
8091e439ec6528a1c50248721a8c68c0c238e31b  CHANGELOG.md
390b735e55a97578f597d0919eedf3e14f43ea07  CHANGELOG-threads.md
```

File: Gemfile.lock [hunk 1]
```diff
@@ -130,7 +159,7 @@ DEPENDENCIES
   yard-junk
 
 RUBY VERSION
-   ruby 2.7.0p0
+   ruby 2.7.1p83
```
[the remark this answers is not in the document]
Comment (ioquatix) [reply Gemfile.lock:162]: I will rebase and fix this issue once #786 is merged. I think bundler just updated the file, it wasn't an explicit decision.

**Comment by ferrarimarco** [comment]
@ioquatix can you rebase against the latest master?

**Comment by ioquatix** [comment]
Rebased.

**Comment by olleolleolle** [comment]
Thanks again, merging.

**Comment by ekohl** [comment]
I'd love to see this in a released version.

**Comment by ioquatix** [comment]
Thanks for your enthusiasm!

**Comment by ioquatix** [comment]
When I made this PR, I didn't know how to efficiently reuse the persistent connections because it was not possible to scope resources to the outer reactor. But now that it is possible using transient tasks, there is some more performance to be had. We don't need to close the persistent internet connection after every block of requests. This is something we should revisit.

**Comment by JacobEvelyn** [comment]
@olleolleolle I'm not sure if you have the ability to do this, but could we get this released in a new RubyGems version? Without this fix the gem is known to produce nondeterministic results (see #774) so it's unusable for me.

**Pull request merged**